src/callbacks/amr_dg2d.jl

```diff
@@ -0,0 +1,330 @@
+                               real(dg), nvariables(equations), polydeg(dg))
+  copy!(cache.interfaces, interfaces)
+
+  # Initialize boundaries
+  boundaries, _ = init_boundaries(leaf_cell_ids, mesh, elements,
+                                  real(dg), nvariables(equations), polydeg(dg))
+  copy!(cache.boundaries, boundaries)
+
+  # Initialize new mortar containers
+  mortars = init_mortars(leaf_cell_ids, mesh, elements,
+                         real(dg), nvariables(equations), polydeg(dg), dg.mortar)
+  copy!(cache.mortars, mortars)
+
+  # Sanity check
+  if isperiodic(mesh.tree) && nmortars(mortars) == 0
+    @assert ninterfaces(interfaces) == 2 * nelements(dg, cache) ("For 2D and periodic domains and conforming elements, the number of interfaces must be twice the number of elements")
+  end
+
+  return nothing
+end
+
+
+# TODO: Taal compare performance of different implementations
+# Coarsen solution data u for four elements, using L2 projection
+function coarsen_elements!(u::AbstractArray{<:Any,4}, element_id, old_u, old_element_id,
+                           adaptor::LobattoLegendreAdaptorL2, equations, dg)
+  @unpack reverse_upper, reverse_lower = adaptor
+
+  # Store old element ids
+  lower_left_id  = old_element_id
+  lower_right_id = old_element_id + 1
+  upper_left_id  = old_element_id + 2
+  upper_right_id = old_element_id + 3
+
+  @boundscheck begin
+    @assert old_element_id >= 1
+    @assert size(old_u, 1) == nvariables(equations)
+    @assert size(old_u, 2) == nnodes(dg)
+    @assert size(old_u, 3) == nnodes(dg)
+    @assert size(old_u, 4) >= old_element_id + 3
+    @assert     element_id >= 1
+    @assert size(    u, 1) == nvariables(equations)
+    @assert size(    u, 2) == nnodes(dg)
+    @assert size(    u, 3) == nnodes(dg)
+    @assert size(    u, 4) >= element_id
+  end
+
+  for j in eachnode(dg), i in eachnode(dg)
+    acc = zero(get_node_vars(u, equations, dg, i, j, element_id))
+
+    # Project from lower left element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, lower_left_id) * reverse_lower[i, k] * reverse_lower[j, l]
+    end
+
+    # Project from lower right element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, lower_right_id) * reverse_upper[i, k] * reverse_lower[j, l]
+    end
+
+    # Project from upper left element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, upper_left_id) * reverse_lower[i, k] * reverse_upper[j, l]
+    end
+
+    # Project from upper right element
+    for l in eachnode(dg), k in eachnode(dg)
+      acc += get_node_vars(old_u, equations, dg, k, l, upper_right_id) * reverse_upper[i, k] * reverse_upper[j, l]
+    end
+
+    # Update value
+    set_node_vars!(u, acc, equations, dg, i, j, element_id)
+  end
+end
+
+
+function indicator_cache(mesh::TreeMesh{2}, equations, dg::DG, cache)
```

Comment:
Currently, we have only one high-level AMR indicator: `IndicatorThreeLevel`. `indicator_cache` is used to create the cache for it, i.e. the final return value of the indicator, which is `< 0` for elements to be coarsened, `> 0` for elements to be refined, and `== 0` for elements which are okay as they are. I am not sure what kind of cache will benecesary for other high-level AMR indicators. Maybe it would be better to use something like `create_cache(::Type{IndicatorThreeLevel}, ...)` instead to be able to use dispatch for other high-level indicators? I tend to think that would indeed be better (and should be documented, of course!). What do you think?

Reply:
Done in 89c1194